packages/caver-transaction/src/transactionTypes/abstractTransaction.js

```diff
@@ -0,0 +1,336 @@
+
+        const err = validateParams(createTxObj)
+        if (err) throw err
+
+        this.gas = createTxObj.gas
+
+        // The variables below are values that the user does not need to pass to the parameter.
+        if (createTxObj.nonce !== undefined) this.nonce = createTxObj.nonce
+        if (createTxObj.gasPrice !== undefined) this.gasPrice = createTxObj.gasPrice
+        if (createTxObj.chainId !== undefined) this.chainId = createTxObj.chainId
+
+        this.signatures = createTxObj.signatures || []
+    }
+
+    /**
+     * @type {string}
+     */
+    get type() {
+        return this._type
+    }
+
+    set type(t) {
+        if (TX_TYPE_STRING[t] === undefined) throw new Error(`Invalid transaction type ${t}. Please refer 'caver.transaction.type'.`)
+        this._type = t
+    }
+
+    /**
+     * @type {string}
+     */
+    get tag() {
+        return this._tag
+    }
+
+    set tag(t) {
+        if (TX_TYPE_TAG[t] === undefined) throw new Error(`Invalid transaction type ${t}. Please refer 'caver.transaction.tag'.`)
+        this._tag = t
+    }
+
+    /**
+     * @type {string}
+     */
+    get nonce() {
+        return this._nonce
+    }
+
+    set nonce(n) {
+        if (!isValidNumber(n)) throw new Error(`Invalid nonce ${n}`)
+        this._nonce = utils.toHex(n)
+    }
+
+    /**
+     * @type {string}
+     */
+    get gas() {
+        return this._gas
+    }
+
+    set gas(g) {
+        if (!isValidNumber(g)) throw new Error(`Invalid gas ${g}`)
+        this._gas = utils.toHex(g)
+    }
+
+    /**
+     * @type {string}
+     */
+    get gasPrice() {
+        return this._gasPrice
+    }
+
+    set gasPrice(g) {
+        if (!isValidNumber(g)) throw new Error(`Invalid gasPrice ${g}`)
+        this._gasPrice = utils.toHex(g)
+    }
+
+    /**
+     * @type {string}
+     */
+    get chainId() {
+        return this._chainId
+    }
+
+    set chainId(ch) {
+        this._chainId = utils.toHex(ch)
+    }
+
+    /**
+     * @type {Array<string>|Array.<Array<string>>}
+     */
+    get signatures() {
+        return this._signatures
+    }
+
+    set signatures(sigs) {
+        this._signatures = refineSignatures(sigs, this.type === TX_TYPE_STRING.TxTypeLegacyTransaction)
+    }
+
+    /**
+     * Signs to transaction with single private key.
+     * @async
+     * @param {Keyring|string} key - The instance of Keyring, private key string or KlaytnWalletKey string.
+     * @param {number} [index] - The index of private key to use.
+     * @param {function} [hasher] - The function to get hash of transaction. In order to use a custom hasher, the index must be defined.
+     * @return {Transaction}
+     */
+    async signWithKey(key, index = 0, hasher = TransactionHasher.getHashForSigning) {
+        // User parameter input cases
+        // (key) / (key index) / (key index hasher)
+        if (_.isFunction(index)) throw new Error(`In order to send a custom hasher as a parameter, the index must be defined first.`)
```

Comment:
Hmm.. I think the second parameter should always be a numeric value. I think it is clear. What do you think?

Reply:
Yes, second parameter(index) should be numeric type.
This logic is only for handling error case that user send custom hasher parameter without defining index.

Reply:
What a kind library this is! :)